vocode/streaming/agent/groq_utils.py

```diff
@@ -0,0 +1,140 @@
+            except IndexError:
+                break
+        if bot_messages_buffer:
+            merged_bot_message = deepcopy(bot_messages_buffer[-1])
+            merged_bot_message.text = " ".join(event_log.text for event_log in bot_messages_buffer)
+            new_event_logs.append(merged_bot_message)
+        else:
+            new_event_logs.append(current_log)
+            idx += 1
+
+    return new_event_logs
+
+def is_phrase_based_action_event_log(event_log: EventLog) -> bool:
+    return (
+        (isinstance(event_log, ActionStart) or isinstance(event_log, ActionFinish))
+        and event_log.action_input is not None
+        and event_log.action_input.action_config is not None
+        and isinstance(
+            event_log.action_input.action_config.action_trigger, PhraseBasedActionTrigger
+        )
+    )
+
+def get_groq_chat_messages_from_transcript(
+    merged_event_logs: List[EventLog],
+    prompt_preamble: str,
+) -> List[dict]:
+    chat_messages = [{"role": "system", "content": prompt_preamble}]
+    for event_log in merged_event_logs:
+        if isinstance(event_log, Message):
+            if len(event_log.text.strip()) == 0:
+                continue
+            else:
+                chat_messages.append(
+                    {
+                        "role": ("assistant" if event_log.sender == Sender.BOT else "user"),
+                        "content": event_log.to_string(include_sender=False),
+                    },
+                )
+        elif isinstance(event_log, ActionStart):
+            action_message: Dict[str, Any]
+            if is_phrase_based_action_event_log(event_log=event_log):
+                pass
+            else:
+                action_message = {
+                    "role": "assistant",
+                    "content": None,
+                    "function_call": {
+                        "name": event_log.action_type,
+                        "arguments": event_log.action_input.params.json(),
+                    },
+                }
+                chat_messages.append(action_message)
+        elif isinstance(event_log, ActionFinish):
+            action_message = {
+                "role": "function",
+                "name": event_log.action_type,
+                "content": event_log.to_string(include_header=False),
+            }
+            chat_messages.append(action_message)
+        elif isinstance(event_log, ConferenceEvent):
+            chat_messages.append(
+                {"role": "user", "content": event_log.to_string(include_sender=False)},
+            )
+    return chat_messages
+
+def format_groq_chat_messages_from_transcript(
+    transcript: Transcript,
+    prompt_preamble: str,
+) -> List[dict]:
+    # merge consecutive bot messages
+    merged_event_logs: List[EventLog] = merge_event_logs(event_logs=transcript.event_logs)
+
+    chat_messages: List[Dict[str, Optional[Any]]]
+    chat_messages = get_groq_chat_messages_from_transcript(
```

Comment:
Since we are deleting the groq version, we will want to import the `openai_utils.py` `get_openai_chat_messages_from_transcript()` method here

Reply:
One question: inside `get_openai_chat_messages_from_transcript()`, we're using some rules of thumb to estimate the number of tokens in the prompt. Do you know if the same rules could be applied to the Groq models? If not, how should we deal with this?

Reply:
Yeah, it needs us to use the tokenizers for the corresponding model, which I don't think groq provides for us in a simple library. For now, we can leave it out but could you add a TODO?

Reply:
Will do, thanks!